### DISCUSSION TITLE
Move to jdk 11

### DISCUSSION THREAD

**Discussion by hazendaz** [revelc/formatter-maven-plugin#517] [General]
This discussion is intended to start discussing if we should move to jdk 11.  I'm actually against this one given how big our community is as it could result in inabiilty for users to upgrade.

The situation currently is that Eclipse moved to jdk 11 with 2020-12.  Not all libraries actually use java 11 as a requirement yet but this is a problem with how we do javascript (jsdt-core).  We have to pull and rebundle that from p2.  As a result, any update post our current point requires osgi with java 11.  The size of that jar we rebundle from class files not source files is around 30mb making it rather hard for us to currently know if there was code tied to java 11 bytecode.  I'm sure there are some tricks we could explore but not sure if its worth it.

Now, from a dev ops standpoint, in my daily work, I'm responsible for well over 500 java builds.  At this time, we are currently north of 85% on jdk 11 or higher builds.  I don't know how typical that is in the community overall and clearly even in my own angle, there is 15% that woudl not benefit if we forced up to jdk 11.

Some plugins have already moved to jdk 11 requirement, not a lot but some have.  We could keep both active as well.  That said, lets discuss this...

**Comment by ctubbsii** [comment]
I think it's reasonable to upgrade to JDK 11, as most developers are using that (or newer) for new Java projects. Existing projects should have no problem continuing to use older versions of this plugin. Even users targeting older JREs can still build their projects in 11 or newer JDK, with the `--release ` flag, so they shouldn't have issues running newer versions of this plugin if it were to require 11 during the Maven build. So, there shouldn't be much reason for this plugin to support running in a Maven environment stuck on pre-11 JDK versions.

FWIW, I checked all the dependencies for JSDT (only checked one class file per distinct package/directory, so it was faster to check the whole 30MB) and there was nothing using anything newer than Java 8. However, that could change. In fact, it's likely to change over time. I think it's reasonable to move forward now, when it's a casual choice, rather than later, when it becomes an urgent one to avoid a breakage.

So, I'd be strongly in favor of bumping to 11. We can communicate the change with a major version bump in the plugin, if it helps. If we find a critical bug in a previous version (to fix an actual bug, not just support a newer Eclipse release) and there's demand for it, we can always patch and release a bugfix for the last release that supported Java 8. If users don't want to use Java 11, they can use the older plugin version. If they do, then they can use the newer plugin version. No big deal.

**Comment by hazendaz** [comment]
Well looks like we are at jdk 11 already...

 Error:  Failed to execute goal net.revelc.code.formatter:formatter-maven-plugin:2.17.0:format (default) on project waffle-jna: Execution default of goal net.revelc.code.formatter:formatter-maven-plugin:2.17.0:format failed: An API incompatibility was encountered while executing net.revelc.code.formatter:formatter-maven-plugin:2.17.0:format: java.lang.UnsupportedClassVersionError: org/eclipse/jdt/core/ToolFactory has been compiled by a more recent version of the Java Runtime (class file version 55.0), this version of the Java Runtime only recognizes class file versions up to 52.0

**Reply by hazendaz** [reply]
We should just now note that for clarity given that class is out of jdt and therefore means the eclipse update itself send us over this edge.  I'll get the jsdt part released again in coming days and just apply that.

**Reply by ctubbsii** [reply]
Ah, I didn't realize JDT had done this, but it makes sense. I think this can be fixed in 2.17.1.

**Reply by ctubbsii** [reply]
Fixed in #533

**Reply by hazendaz** [reply]
How should we close these out?  They seem useful for others but there is no way to do much other than delete it.  Maybe a new label completed or implemented?

**Reply by ctubbsii** [reply]
Just ignore it? If you want to label it as "archived", or similar, fine, but there's no reason to do that. If this were a mailing list, nobody would be marking old mailing list threads that way. Any conversation can be revisited.

**Reply by hazendaz** [reply]
ok I think I will add a label of 'archived' as you noted.  Its more for me when I'm not here for a while then trying to see if anything needs looked at.  Given we addressed this one its clear and emails I receive as well so I would know if anyone asked for something new but some are truly archived - well outside of questions of why and what not.

**Reply by ctubbsii** [reply]
If it's just for you, GitHub has notifications, that are organized per-repository. You can just look at your notifications inbox in GitHub's UI.

**Reply by hazendaz** [reply]
I get 100+ emails a day from github.  I'm working on something like 100+ projects here and 800+ at work.  So its more for me in that regard so when I come back as I will certainly delete any emails that I'm either not going to answer or was just dependabot or renovate work.  Plus its too hard to keep track that far back.  The feature here looks great with archive ad it lists on the first page so no need to even peak in.  I'm sure github will improve this over time.  A closed option would have been good :)

**Comment by hazendaz** [comment]
we are on jdk 11 now due to underlying Eclipse forcing us up.
